### DISCUSSION TITLE
What to print for "gplately.__version__"?

### DISCUSSION THREAD

**Discussion by jcannon-gplates** [GPlates/gplately#263] [Ideas]
My development version of gplately prints the PyPI release version, instead of the development version, with `gplates.__version__`.

Currently it appears that, between official releases, we need to set [this](https://github.com/GPlates/gplately/blob/e15e361b60f1b97645735d503df0dcdef4cd19d6/setup.py#L39) to None so that the [git version](https://github.com/GPlates/gplately/blob/e15e361b60f1b97645735d503df0dcdef4cd19d6/setup.py#L70) is used?

However, there's also a version [here](https://github.com/GPlates/gplately/blob/e15e361b60f1b97645735d503df0dcdef4cd19d6/gplately/__init__.py#L203) (in the `__init__.py`), which is referenced by `pyproject.toml` [here](https://github.com/GPlates/gplately/blob/e15e361b60f1b97645735d503df0dcdef4cd19d6/pyproject.toml#L46).

PyGPlates uses PEP 440 versioning which has the `.dev` suffix - see [here](https://www.gplates.org/docs/pygplates/generated/pygplates.version).

So maybe gplately can also use PEP 440?  And maybe we can update the development version automatically (eg, when merging a pull request).

By the way, the new pyproject.toml in pyGPlates (to be released soon) extracts the version from a specific file (so the version only needs to be changed in one place):

```
#
# Extract version from 'PYGPLATES_PEP440_VERSION' variable in 'cmake/modules/Version.cmake'.
#
[tool.scikit-build.metadata.version]
provider = "scikit_build_core.metadata.regex"
input = "cmake/modules/Version.cmake"
regex = 'set\(\s*PYGPLATES_PEP440_VERSION\s*(?P<value>\S+)\s*\)'
```

...however, for gplately, the current approach seems fine (`version = {attr = "gplately.__version__"}`). I guess I had to do it differently in PyGPlates because it's a C++ extension (not pure Python).

**Comment by jcannon-gplates** [comment]
Also, I know this was addressed in #253, but maybe once we have dev versions that are different than the official PyPI release version we can then remove the warning [here](https://github.com/GPlates/gplately/blob/e15e361b60f1b97645735d503df0dcdef4cd19d6/gplately/__init__.py#L214-L230)? It's great that you can currently disable it with an environment variable, but I have to remember to set it, or to change `__init__.py`.

**Reply by michaelchin** [reply]
> Also, I know this was addressed in #253, but maybe once we have dev versions that are different than the official PyPI release version we can then remove the warning [here](https://github.com/GPlates/gplately/blob/e15e361b60f1b97645735d503df0dcdef4cd19d6/gplately/__init__.py#L214-L230)? It's great that you can currently disable it with an environment variable, but I have to remember to set it, or to change `__init__.py`.

I think the warning message is useful. It sends users a clear message that they are using unstable code. It is a very useful reminder for me.

**Comment by jcannon-gplates** [comment]
Bonus points for specifying which version a function/class was added/modified (in the API documentation).

I currently do this in pyGPlates manually. For example, I'll add a new parameter to a function and manually specify which version this happened in the documentation, for example, it says *Changed in version 0.26: anchor_plate_id no longer defaults to zero* in [pygplates.RotationModel.get_rotation()](https://www.gplates.org/docs/pygplates/generated/pygplates.rotationmodel#pygplates.RotationModel.get_rotation).

Note that I'm still using pre-1.0 versions, so I don't yet explicity use the PEP440 `.dev` suffix. However, once pygplates reaches `1.0` I will start using it (and incrementing it every time something changes the API).

**Comment by michaelchin** [comment]
we can use setuptools-git-versioning https://github.com/GPlates/gplately/pull/262/commits/205872cdb774d2fca2547908d1a69eb42d433596

**Reply by jcannon-gplates** [reply]
Nice👍. So this now essentially uses [PYPI_VERSION](https://github.com/GPlates/gplately/blob/e15e361b60f1b97645735d503df0dcdef4cd19d6/setup.py#L84), right?

**Reply by michaelchin** [reply]
> Nice👍. So this now essentially uses [PYPI_VERSION](https://github.com/GPlates/gplately/blob/e15e361b60f1b97645735d503df0dcdef4cd19d6/setup.py#L84), right?

Yes, the Python code in __init__.py will read the gplately package metadata to get the version number so that the pip package version is always the same with the gplately.__version__.

This is just a proposal from me. There are other ways to do it. I will just leave it there until we come up with a smarter solution.

**Reply by jcannon-gplates** [reply]
I was just looking again at how the versioning is done. I like your setuptools-git-versioning approach.

What I don't yet understand is how `pyproject.toml` and `setup.py` interact?

It seems both can co-exist, with `pyproject.toml` being the [more modern approach](https://packaging.python.org/en/latest/guides/modernize-setup-py-project/#modernize-setup-py-project) (eg, can specify the build back-end).

In our case there's seems to be a lot of overlap, where parameters are specified in both `pyproject.toml` and `setup.py`. Like dependencies, URLs, descriptions, etc. In fact, the only thing in `setup.py` that's not in `pyproject.toml` is the version: https://github.com/GPlates/gplately/blob/b2ba5aecac851aa226e0cd059647c91b32082ed0/setup.py#L37-L70

...and I'm not sure that `git_version()` is even used (since `PYPI_VERSION` is never set to `None`, I think). And your use of setuptools-git-versioning essentially replaces all that, I think.

I'm guessing that setuptools-git-versioning somehow finds PYPI_VERSION (in `setup.py`): https://github.com/GPlates/gplately/blob/b2ba5aecac851aa226e0cd059647c91b32082ed0/gplately/utils/version.py#L17-L26

...but I'm not sure exactly how.

In any case, do you think we should just get rid of `setup.py` to avoid confusion with `pyproject.toml`? This would include moving the version into `pyproject.toml` (where setuptools-git-versioning would presumably still find it).

**Reply by michaelchin** [reply]
> What I don't yet understand is how `pyproject.toml` and `setup.py` interact?

I don't think they interact. The `pyproject.toml` should just work without `setup.py`.

**Reply by michaelchin** [reply]
> do you think we should just get rid of `setup.py` to avoid confusion with `pyproject.toml`?

I suggested removing `setup.py` before. I vaguely recall someone would like to keep `setup.py` for some reason. I cannot remember the details. So we decided to keep `setup.py` for a while longer. I am totally happy to get rid of `setup.py`.

**Reply by michaelchin** [reply]
> This would include moving the version into `pyproject.toml` (where setuptools-git-versioning would presumably still find it).

I am not sure about this. My understanding is that the `setuptools-git-versioning` gets the version info from `git`. I am not sure the meaning of "moving the version into `pyproject.toml`".

**Reply by jcannon-gplates** [reply]
> > What I don't yet understand is how pyproject.toml and setup.py interact?
>
> I don't think they interact. The pyproject.toml should just work without setup.py.

Yeah, that's what I thought too. Was just wondering why we have both currently, but you answered that next...

> I suggested removing setup.py before. I vaguely recall someone would like to keep setup.py for some reason. I cannot remember the details. So we decided to keep setup.py for a while longer. I am totally happy to get rid of setup.py.

It might have been mentioned by Ben in #140. In particular...

> Apparently `pyproject.toml` files are going to become the new standard for python packages. Currently `setup.py` is still [recommended to stay](https://packaging.python.org/en/latest/guides/modernize-setup-py-project/), but there are some advantages to the new packaging system - build time dependencies in particular!

It looks like the `recommended to stay` link just says you don't need to delete `setup.py`. And it points to [this](https://packaging.python.org/en/latest/discussions/setup-py-deprecated/#setup-py-deprecated) link which says `setup.py` can still be used to configure parameters (similar to `pyproject.toml`) but we just shouldn't use it for things like `python setup.py install` anymore (since that's deprecated in favour of `pip install .`).

So, yeah, I think we can remove `setup.py`. As far as I can tell, it doesn't do anything that our `pyproject.toml` doesn't do, except set `PYPI_VERSION`, which brings me to...

> I am not sure the meaning of "moving the version into pyproject.toml".

...I just meant the `PYPI_VERSION` in `setup.py`, because I thought that's where the version is ultimately coming from somehow.

> > This would include moving the version into pyproject.toml (where setuptools-git-versioning would presumably still find it).
>
> I am not sure about this. My understanding is that the setuptools-git-versioning gets the version info from git.

Ok, but I don't know where git gets the version from?

When I do a search for "1.3" in GitHub, I only see it in `PYPI_VERSION` (in `setup.py`).

Or maybe it's getting it from a "tag" in git?  But the code snippet (that I posted above) uses `importlib.metadata.version()`, and the comment says "get the version string from GPlately package metadata".

So it sounds like it's getting the version from some kind of package metadata, but I don't know where or how?  And that's pretty much where the trail went cold for me.  If you have any answers that'd be great?

But yeah, if we can just remove `setup.py` and forget about `PYPI_VERSION` then that would be great!  No point having to update the same parameters (like dependencies, URLs, descriptions, etc) in two different places (not knowing which ones are actually being read during packaging).

**Reply by michaelchin** [reply]
> If you have any answers that'd be great?

If my memory serves me right, I think the version info was from git tag when building the package. And then the version info was embedded somehow into the package's metadata. At run time, I think GPlately will get the version info from package's metadata. But the version info was originated from git tag in the first place.

**Reply by jcannon-gplates** [reply]
Ok thanks, that makes sense.

> I think the version info was from git tag

Ah yes, it does look like the git tag is being used (as described [here](https://setuptools-git-versioning.readthedocs.io/en/stable/schemas/tag/tag_release.html)). And explains why we get versions like `1.0.0.post1+git.64e68cd4` after a release.

That's good - it means we can ignore `PYPI_VERSION` in `setup.py`.

> At run time, I think GPlately will get the version info from package's metadata. But the version info was originated from git tag in the first place.

Ok, yes, that would explain why we essentially do this:

```
__version__ = importlib.metadata.version("gplately")
```

It all makes sense now! Thanks.

**Reply by jcannon-gplates** [reply]
So let's just remove `setup.py` for the `2.0.0` release. Shall I create an issue?

**Reply by michaelchin** [reply]
> So let's just remove `setup.py` for the `2.0.0` release. Shall I create an issue?

Yes. Sure, please.

**Reply by jcannon-gplates** [reply]
Great, I've create #323.
